cuda_core/cuda/core/experimental/_device.py

```diff
@@ -957,26 +957,34 @@ def __new__(cls, device_id=None):
 
         # important: creating a Device instance does not initialize the GPU!
         if device_id is None:
-            device_id = handle_return(runtime.cudaGetDevice())
-            assert_type(device_id, int)
+            err, dev = driver.cuCtxGetDevice()
+            if err == 0:
+                device_id = int(dev)
+            else:
+                ctx = handle_return(driver.cuCtxGetCurrent())
+                assert int(ctx) == 0
+                device_id = 0  # cudart behavior
         else:
-            total = handle_return(runtime.cudaGetDeviceCount())
-            assert_type(device_id, int)
-            if not (0 <= device_id < total):
+            total = handle_return(driver.cuDeviceGetCount())
```

Comment:
Assuming that the happy path is common, this `driver.cuDeviceGetCount()` call seems redundant.

Also assuming it's not actually worth the cycles checking for `isinstance`, we could replace the `else` block here with:

```python
        elif device_id < 0:
            raise ValueError(f"device_id must be >= 0, got {device_id!r}")
```

Then below (new line 998) we could do this:

```python
        try:
            return devices[device_id]
        except IndexError:
            raise ValueError(f"device_id must be within [0, {len(devices)}), got {device_id!r}")
```

WDYT?

Reply:
I think this is an excellent idea, done in commit 6b245ff17. It also optimized the case with an explicit device ID by almost 4x, see the updated PR description.